src/main/ruby/core/dir.rb

```diff
@@ -49,22 +53,53 @@ def initialize(path, options=undefined)
       enc = Rubinius::Type.coerce_to_encoding enc if enc
     end
 
-    Truffle.invoke_primitive :dir_open, self, path, enc
+    @encoding = enc || Encoding.find('filesystem')
+
+    @ptr = Truffle::POSIX.opendir(@path)
+    Errno.handle if @ptr.null?
+  end
+
+  private def ensure_open
```

Comment:
I think this would look a bit better as `ensure_open!`

Reply:
It's consistent with `IO#ensure_open` and I think `!` should only be used if there is a `!`-less variant in MRI naming conventions.
But of course this is our internal API, so I can change if you want.

Reply:
The other convention I've seen is to do it if there's a side effect or exception raised. I think that's mostly done in Rails. I don't feel strongly enough about it though.